### ISSUE TITLE
[Request] Add support for prowlarr

### ISSUE THREAD

**Issue by jaja360** [calibrain/calibre-web-automated-book-downloader#28]
I think many people who are currently using `calibre + calibre-web + readarr` to manage / download their books are looking at `calibre-web-automated + calibre-web-automated-book-downloader` as a more and more desirable alternative.

The only thing missing to become a complete replacement I believe would be an integration with prowlarr, so that if the requested books aren't found directly by `calibre-web-automated-book-downloader` (you guys need a shorter name for the app, lol), a request to the indexers in prowlarr could be made.

**Comment by noseboy** [comment]
Yeah, rename it to cwabd!

**Comment by yroyathon** [comment]
I've been calling it "ebookseerr" :) .

**Comment by calibrain** [comment]
From my understanding, Prowlarr does not support direct download, but only nzb and torrents
I have never used it myself, so I am not sure how much work to add support for it. 
But if someone is more familiar with it and wants to help, I would appreciate it !

(for the name, I am just calling it to `cwa-bd`)

**Comment by calibrain** [comment]
I am not sure to understand
What would be the workflow here ?

User goes on Readarr, requests a book -> prowlarr gets request transfers to cwa-bd -> cwa-bd downloads it, and adds a fake completed nzb -> prowlarr thinks the nzb was downloaded successfully and now the book is available -> readarr is happy

And in the backend cwa automatically ingested the new book >

**Comment by W3Max** [comment]
Hi @calibrain! 

Here is what I think people would like, but...
  1 - It is a lot of work!
  2 - Maybe I do not understand what people would like...

Your tool is currently a "book searcher" and a "download client".

I think people would like to get rid of Readarr. It is complex and it might not be maintained anymore. Your tool is simple and just works.

People using Prowlarr (*arr apps) already have "download clients" configured for torrents and/or Usenet (like SABnzbd).

Sonarr/Readarr/Radarr are not download clients. They are searchers, ie, you need to configure one or more download clients for them to do something useful.

As for Prowlarr, it is only an app that allows to centralize the configurations (like indexers) for the "download clients", so that you don't have to do it in each *arr apps.

Basically, I think what people are saying is that if it was possible to configure our own "download clients" inside calibre-web-automated-book-downloader, and use it only for searching/queuing downloads, it could be nice.

But your tool is great as it is and I thank you for that!

**Comment by kylehandy123** [comment]
@calibrain May I suggest merging issue #21 into this issue? My feature request ties into Prowlarr as well, and the overall request of this issue.
Thanks!

**Comment by calibrain** [comment]
Yeah, this is complex and if I am being honest, out of scope of what I am trying to do with cwa-bd
But I understand the need for this, and if I have some free time, I can try working on this, but I can't make any promise.
But if anyone wants to work on this feature, I will help as much as I can !

**Comment by cathelijne** [comment]
Might be worth it to look at https://github.com/riffsphereha/newznabarr, a plugin for prowlarr that mimics sabnzbd and currently downloads from libgen. Adding Anna's to that would be awesome, and adding libgen to CWABD would _also_ be awesome :)

That being said, in the comments above it's already mentioned that searchers and downloaders are different beasts, and I tend to agree. Just thought it was worthwhile to mention another cool search/download alternative.

**Comment by mackadoo** [comment]
I figure I would add here rather than create a new similar issue - adding debrid support would be awesome. Currently Torbox has added Anna's Archive as a supported file hoster and just sending the url of a book on AA to the debrid web downloader will pull from AA into Torbox and then the user can download the file from there, bypassing cloudflare checks and wait times. Any book I have tried I can download in 30sec or so this way.

**Comment by alexhb1** [comment]
I might be able to take a stab at this @calibrain if you're up for it? Bit of a bigger addition but hopefully one that can be added largely separate and not interfere with the existing AA downloader. It likely wouldn't be a full featured readarr / *arr type experience, instead just offering an alternative search provider and implementations with prowlarr/download clients, and using the same UI. It would also be surprisingly little amounts of code to add, features to build etc. Way less complicated than AA downloading and instead mostly just calls to prowlarr / download clients and a nice UI to show it. 

I can start making a plan of how it could look :)

**Comment by calibrain** [comment]
I wonder if it could just be a simple prowlarr plugin ? Since CWA-BD can be used fully through the API ?
But sure, if you want to try something for it, I am all good with it !
I prefer keeping things as simple as possible, but if it can be in its own folder ( a bit like the custom browser we use, its in its own folder and its own logic and can be easily disabled) I am good with it !

**Comment by alexhb1** [comment]
Yep that's hopefully the plan. The existing code might need a bit of a restructure just to separate out the AA side of things from the basic functionality, and then it's quite straightforward to drop-in the prowlarr side without ever affecting the existing downloader :)

A few key parts probably: Searching via book metadata (choose your provider and API key), a page to view prowlarr search results, and updating the download sidebar to show external download client progress. The rest is just various calls to the prowlarr API and talking to the download clients.

**Comment by gravelfreeman** [comment]
> I might be able to take a stab at this [@calibrain](https://github.com/calibrain) if you're up for it? Bit of a bigger addition but hopefully one that can be added largely separate and not interfere with the existing AA downloader. It likely wouldn't be a full featured readarr / *arr type experience, instead just offering an alternative search provider and implementations with prowlarr/download clients, and using the same UI. It would also be surprisingly little amounts of code to add, features to build etc. Way less complicated than AA downloading and instead mostly just calls to prowlarr / download clients and a nice UI to show it.
> 
> I can start making a plan of how it could look :)

You might want not to invest time in CWA at all. I've read complaints about ignored PRs, security issues, etc.

People seems to be moving towards [Booklore](https://github.com/booklore-app/booklore) + [Ephemera](https://github.com/OrwellianEpilogue/ephemera) combo.

[Booklore](https://github.com/booklore-app/booklore) is finally the long awaited alternative to replace Calibre completely. Let's be honest Calibre is outdated.

[Ephemera](https://github.com/OrwellianEpilogue/ephemera) is like [Openbooks](https://github.com/evan-buss/openbooks) and CWAD but can be used in Readarr. Although Readarr went public archive a few months ago and users have been replacing it with [rreading-glasses](https://github.com/blampe/rreading-glasses).

We really need an application to download books from any source :

- IRC
- Torrent
- Usenet
- DDL?

Maybe we should start seeking interest at Ephemera, rreading-glasses or do a completely different application that does it all once and for all.

**Comment by alexhb1** [comment]
@gravelfreeman This app has nothing to do with CWA other than the name really, use it however you like. I use it with booklore for example. 

It's just a nice and simple downloader and more user-friendly than others. My parents save it as a PWA on their phone and send books to their kindles all the time for example. It's also pretty plug-and-play, install the main version and you're away with AA slow downloads with little to no setup. 

As long as it's implemented flexibly, prowlarr would be excellent to use in this app alongside AA as a power-user option. Any thing else like IRC hopefully can be added in a similar easy way down the line.

**Comment by alexhb1** [comment]
I'll be working on this at https://github.com/calibrain/calibre-web-automated-book-downloader/discussions/346 :)

**Comment by yroyathon** [comment]
Readarr is now dead, so fitting exactly into that ecosystem isn't really useful anymore.  

My workflow starts by using CWABD to search for/download a book.  99.9% of the time that is it, it finds it, it places the book into CWA, done.  I can't remember the last time it couldn't find a book.  However in that case, I do a manual search in prowlarr (for ebooks, that's basically MaM finding it, or rarely some usenet indexer if not MaM), and post processing scripts place it in CWA.  So for me, adding prowlarr probably wouldn't add much value.  But maybe the books I'm searching for are vanilla compared to what everyone else is searching for, and Anna's is often not successful in finding the books for other folks.

I often think about the idea of smaller apps doing one thing well, vs larger apps doing many things poorly.  So far, CWABD is the former, it's great.  It sounds like a few of you are dedicated to preventing the 2nd from happening.  Good luck adding this functionality!

**Comment by alexhb1** [comment]
> Readarr is now dead, so fitting exactly into that ecosystem isn't really useful anymore.
> 
> My workflow starts by using CWABD to search for/download a book. 99.9% of the time that is it, it finds it, it places the book into CWA, done. I can't remember the last time it couldn't find a book. However in that case, I do a manual search in prowlarr (for ebooks, that's basically MaM finding it, or rarely some usenet indexer if not MaM), and post processing scripts place it in CWA. So for me, adding prowlarr probably wouldn't add much value. But maybe the books I'm searching for are vanilla compared to what everyone else is searching for, and Anna's is often not successful in finding the books for other folks.
> 
> I often think about the idea of smaller apps doing one thing well, vs larger apps doing many things poorly. So far, CWABD is the former, it's great. It sounds like a few of you are dedicated to preventing the 2nd from happening. Good luck adding this functionality!

Thanks, and yeah. I don't think trying to be like readarr is useful at all. Plus the arr apps are a real chore in my opinion 🤣. 

This app is a search engine for books, looks like one, and I'd like it to stay super simple out of the box (My parents got using it with zero effort or guidance, that's a genuine miracle lol). If we can add more sources and configuration options on the backend without making the core app stray away from being an easy search engine for books, then that's pretty solid to me.

**Comment by gravelfreeman** [comment]
Sorry I thought this was some kind of plugin to CWA. Then this is very good news! I just checked the yaml and I guess I can just point the download directory to the Booklore ingest folder. 

As for the use case of torrents, Usenet and ddl sites, I agree that irc highway has almost everything in English. I’m French and irc highway doesn’t have as many books as French trackers and French ddl private sites.

So I’m still vouching for an aio downloader for “whatever book app you’re using”. 

As for readarr I agree it’s a lot of overhead and it’s time consuming when the only advantage I see that it can download a book on your wantlist when it gets uploaded on your indexers.

**Comment by alexhb1** [comment]
Hopefully with the backend work I'm doing it would be relatively easy to drop in an irc_highway.py file for example, and have an extra source added just like that. The UI and system will be abstracted enough that as long as you have the download functionality worked out for your source - the UI, the scheduler, etc, will just work with whatever source you throw at it. Might come close to an all-in-one app for this sort of thing :)

**Comment by alexhb1** [comment]
WIP preview of how it's looking so far!

<img width="1350" height="728" alt="Image" src="https://github.com/user-attachments/assets/6638c8da-630d-4cce-af08-5fb057720f3f" />

**Comment by alexhb1** [comment]
Plugin system is now in place and search via Hardcover / Open Library added in the `dev` branch. Prowlarr is functioning on my local build but just need to do some more testing before this gets committed :)

**Comment by alexhb1** [comment]
Update now available on the dev branch for testing - use `calibre-web-automated-book-downloader:dev` in your compose to update. Full changes at #361 :)

Quick summary is that Prowlarr and IRC are now available as new sources within the app's universal search, same as Anna's Archive. Configure a metadata provider, search your books, and view book downloads from all three sources in a single place. 

Key bits: 

- Universal search mode needs to be enabled and a metadata provider set (Hardcover is reccomended for best results)
- Prowlarr works with torrent and usenet. Select specific trackers to search in settings, or set to search all. 
- Configure your download client for torrent/NZB, either Qbittorrent/Deluge/Transmission, and NZBget/SABnzbd. 
- Full download status and live progress in the UI sidebar, same as IRC and Anna's Archive downloads. All search preferences and file processing is done at the app level, same for all three sources. 
- IRC is entirely self contained, no download client required. If you've used OpenBooks before the concept will look very similar. 

I've done a lot of testing but naturally there's a lot there that's hard to do solo, so please give it a test and submit issues for anything you find. I've also not got access to MAM so anyone with access there could let me know how the experience is with that would be really appreciated :)

**Comment by Xytronix** [comment]
Thats great would you add Undernet?

irc.undernet.org #bookz channel

**Comment by pequesan** [comment]
I know this may sound a little bit specific and "tricky", but here in Spain we still use emule, and it works really well as a epub book download source.

I won't ask you to develop a complete interface with *mule clients... because 90% it's done and its name is Amarr:

https://github.com/vexdev/amarr

Amarr interfaces between the "amule" client and the *arr software set. It exposes Amule search as a "torznab indexer", and Amule download as "qbittorrent download client". I'm sucessfully using it with "Readarr".

I've updated to CWADB dev branch to test this. I've been able to add my Amarr as a torznab indexer in Prowlar, and my Prowar to CWABD. CWABD sees Amarr indexer and I can choose it in CWABD settings. But I have 2 problems:

- Searches dont't arrive to Prowlarr. At least there's no Prowlarr log entry. CWADB log shows 0 results:

2025-12-30 22:52:03,530 - cwa_book_downloader.release_sources.prowlarr.source - DEBUG - source.py:372 - Prowlarr: no results found

- I fail trying to add my Amarr as qbittorrent download client in CWABD. In any *arr software, qbittorrent download client is added as host and port (two separate fields) while in CWABD you must specify the url... but "http://host:port" doesn't work. I've got qbittorrent running in my server and I've test to put its url, and it works. So this has something to do with the specific Amarr qbittorent emulation implementation.

Could you take a look?

**Comment by alexhb1** [comment]
Now released :)

**Comment by tchirou** [comment]
Hi. Would it be possible for prowlarr to add as a download client a blackhole folder for torrent files ? I only have rdtclient which I seem to get working. Thanks a lot
